lib/block_view.go

```diff
@@ -7060,6 +7342,153 @@ func (bav *UtxoView) _connectSwapIdentity(
+		return 0, 0, nil, fmt.Errorf("_connectAuthorizeDerivedKey: called with bad OperationType %s",
+			txn.TxnMeta.GetTxnType().String())
+	}
+
+	// Make sure transaction hasn't expired.
+	if txMeta.ExpirationBlock <= uint64(blockHeight) {
+		return 0, 0, nil, RuleErrorAuthorizeDerivedKeyExpiredDerivedPublicKey
+	}
+
+	// Validate the owner public key.
+	ownerPublicKey := txn.PublicKey
+	if len(ownerPublicKey) != btcec.PubKeyBytesLenCompressed {
+		return 0, 0, nil, RuleErrorAuthorizeDerivedKeyInvalidOwnerPublicKey
+	}
+	if _, err := btcec.ParsePubKey(ownerPublicKey, btcec.S256()); err != nil {
+		return 0, 0, nil, errors.Wrap(RuleErrorAuthorizeDerivedKeyInvalidOwnerPublicKey, err.Error())
+	}
+
+	// Validate the derived public key.
+	derivedPublicKey := txMeta.DerivedPublicKey
+	if len(derivedPublicKey) != btcec.PubKeyBytesLenCompressed {
+		return 0, 0, nil, RuleErrorAuthorizeDerivedKeyInvalidDerivedPublicKey
+	}
+	if _, err := btcec.ParsePubKey(derivedPublicKey, btcec.S256()); err != nil {
+		return 0, 0, nil, errors.Wrap(RuleErrorAuthorizeDerivedKeyInvalidDerivedPublicKey, err.Error())
+	}
+
+	// Verify that the access signature is valid. This means the derived key is authorized.
+	err := _verifyAccessSignature(ownerPublicKey, derivedPublicKey,
+		txMeta.ExpirationBlock, txMeta.AccessSignature)
+	if err != nil {
+		return 0, 0, nil, errors.Wrap(RuleErrorAuthorizeDerivedKeyAccessSignatureNotValid, err.Error())
+	}
+
+	// Get current (previous) derived key entry. We might revert to it later so we copy it.
+	prevDerivedKeyEntry := bav._getDerivedKeyMappingForOwner(ownerPublicKey, derivedPublicKey)
+
+	// We make sure that the key hasn't been invalidated in the past. If that's the case,
+	// we return. A derived key can't be re-authorized after being revoked in the past.
+	if prevDerivedKeyEntry != nil && !prevDerivedKeyEntry.isDeleted {
+		if prevDerivedKeyEntry.OperationType == AuthorizeDerivedKeyOperationNotValid {
+			return 0, 0, nil, RuleErrorAuthorizeDerivedKeyDeletedDerivedPublicKey
+		}
+	}
+
+	// At this point we've verified the access signature, which means the derived key is authorized
+	// to sign on behalf of the owner. We now add a temporary derived key to utxoView with
+	// OperationType set to Valid so we pass txn signature verification in _connectBasicTransfer.
+	derivedKeyEntry := DerivedKeyEntry {
+		*NewPublicKey(ownerPublicKey),
+		*NewPublicKey(derivedPublicKey),
+		txMeta.ExpirationBlock,
+		AuthorizeDerivedKeyOperationValid,
+		false,
+	}
+	bav._setDerivedKeyMapping(&derivedKeyEntry)
+
+	// Call _connectBasicTransfer to verify txn signature.
+	totalInput, totalOutput, utxoOpsForTxn, err := bav._connectBasicTransfer(
+		txn, txHash, blockHeight, verifySignatures)
+	if err != nil {
+		// Since we've failed, we revert the utxoView mapping to what it was previously.
+		// We're doing this manually because we haven't fully connected the txn yet.
+		bav._deleteDerivedKeyMapping(&derivedKeyEntry)
```

Comment:
Heh. Right now we don't clean up if an error occurs. Rather, the caller is supposed to assume the view is completely broken when this happens. Just to keep things consistent, let's revert these two lines and just return the error. Same for below.

Reply:
Oh I see. This is a little hackey what you're doing here. Hmm...

Reply:
I'm confused. Why do you need to setDerivedKeyMappings above before calling _connectBasicTransfer? It seems like it should work without you having to do this. Are you trying to support the case where a derived key can create a sub-derived-key?

Reply:
I'm allowing the Authorize transaction to be signed by the derived key (since the access signature is already signed by the owner). For first-time keys there's no existing entry in the DB so `_verifySignature()` would fail. If I add the key before calling the `_connectBasicTransfer` then it's loaded in utxoview and `_verifySignature()` passes.

Reply:
I think you can still revert the delete+set. If it errors then the view is assumed to be compromised and we'll generate a new one.

Reply:
Eh.. on second thought just keep the _delete+_set. Doesn't hurt.